File: pkg-r/R/registry.R
# Load source definitions from the vendored shared spec.

.biogate_cache <- new.env(parent = emptyenv())

.load_sources <- function() {
  if (!is.null(.biogate_cache$sources)) {
    return(.biogate_cache$sources)
  }
  dir <- system.file("extdata", "sources", package = "biogate")
  files <- sort(list.files(dir, pattern = "\\.yaml$", full.names = TRUE))
  reg <- list()
  for (f in files) {
    spec <- yaml::read_yaml(f)
    reg[[spec$key]] <- spec
  }
  .biogate_cache$sources <- reg
  reg
}

#' List available source databases
#'
#' @return A character vector of source keys, sorted.
#' @seealso [source_info()] for a table with more detail.
#' @examples
#' sources()
#' @export
sources <- function() {
  sort(names(.load_sources()))
}

.source_modes <- function(spec) {
  modes <- "pattern"
  if (!is.null(spec$cache)) {
    modes <- c(modes, "cache")
  }
  if (!is.null(spec$remote)) {
    modes <- c(modes, "remote")
  }
  paste(modes, collapse = ", ")
}

#' Describe the available sources
#'
#' Answers "what does a valid id look like and how can I check it?" for each
#' source.
#'
#' @return A [tibble][tibble::tibble] with one row per source and the columns
#'   `key`, `name`, `example` (a valid identifier for the source), `modes` (the
#'   checking modes it supports), `species_aware`, and `version_aware`.
#' @examples
#' source_info()
#' @export
source_info <- function() {
  reg <- .load_sources()
  keys <- sort(names(reg))
  tibble::tibble(
    key = keys,
    name = vapply(
      keys,
      function(k) reg[[k]]$name,
      character(1),
      USE.NAMES = FALSE
    ),
    example = vapply(
      keys,
      function(k) {
        ex <- reg[[k]]$example
        if (is.null(ex)) NA_character_ else ex
      },
      character(1),
      USE.NAMES = FALSE
    ),
    modes = vapply(
      keys,
      function(k) .source_modes(reg[[k]]),
      character(1),
      USE.NAMES = FALSE
    ),
    species_aware = vapply(
      keys,
      function(k) isTRUE(reg[[k]]$species_aware),
      logical(1),
      USE.NAMES = FALSE
    ),
    version_aware = vapply(
      keys,
      function(k) isTRUE(reg[[k]]$version_aware),
      logical(1),
      USE.NAMES = FALSE
    )
  )
}

.get_source <- function(source_db) {
  reg <- .load_sources()
  if (!source_db %in% names(reg)) {
    cli::cli_abort(
      c(
        "Unknown {.arg source_db} {.val {source_db}}.",
        i = "Available sources: {.val {sources()}}."
      ),
      class = "biogate_error_unknown_source"
    )
  }
  reg[[source_db]]
}

File: pkg-py/src/biobouncer/_registry.py
"""Load source definitions from the vendored shared spec."""

from __future__ import annotations

import functools
from dataclasses import dataclass
from importlib.resources import files

import yaml


class UnknownSourceError(ValueError):
    """Raised when a source key is not in the registry."""


@dataclass(frozen=True)
class Source:
    key: str
    name: str
    description: str
    pattern: str
    species_aware: bool
    version_aware: bool
    curie: dict | None
    normalize: dict | None
    cache: dict | None
    remote: dict | None
    species: dict | None = None
    example: str | None = None
    default_version: str | None = None
    suggest: dict | None = None

    def modes(self) -> list[str]:
        """Return the checking modes this source supports, in order."""
        out = ["pattern"]
        if self.cache:
            out.append("cache")
        if self.remote:
            out.append("remote")
        return out


@functools.lru_cache(maxsize=1)
def _registry() -> dict[str, Source]:
    root = files("biobouncer") / "_data" / "sources"
    out: dict[str, Source] = {}
    for entry in sorted(root.iterdir(), key=lambda p: p.name):
        if not entry.name.endswith(".yaml"):
            continue
        spec = yaml.safe_load(entry.read_text(encoding="utf-8"))
        out[spec["key"]] = Source(
            key=spec["key"],
            name=spec.get("name", spec["key"]),
            description=spec.get("description", ""),
            pattern=spec["pattern"],
            species_aware=bool(spec.get("species_aware", False)),
            version_aware=bool(spec.get("version_aware", False)),
            curie=spec.get("curie"),
            normalize=spec.get("normalize"),
            cache=spec.get("cache"),
            remote=spec.get("remote"),
            species=spec.get("species"),
            example=spec.get("example"),
            default_version=spec.get("default_version"),
            suggest=spec.get("suggest"),
        )
    return out


def sources() -> list[str]:
    """Return the sorted list of available source keys."""
    return sorted(_registry())


def source_info() -> list[dict]:
    """Describe every source, one dict per source, sorted by key.

    Each dict answers "what does a valid id look like and how can I check it?".

    Returns:
        A list of dicts with keys ``key``, ``name``, ``example`` (a valid
        identifier for the source), ``modes`` (the checking modes it supports),
        ``species_aware``, and ``version_aware``.

    Example:
        >>> import biobouncer as bg
        >>> info = {row["key"]: row for row in bg.source_info()}
        >>> info["mondo"]["example"]
        'MONDO:0005148'
        >>> info["mondo"]["modes"]
        ['pattern', 'cache', 'remote']
    """
    reg = _registry()
    return [
        {
            "key": src.key,
            "name": src.name,
            "example": src.example,
            "modes": src.modes(),
            "species_aware": src.species_aware,
            "version_aware": src.version_aware,
        }
        for src in (reg[k] for k in sorted(reg))
    ]


def get_source(key: str) -> Source:
    """Return the Source for ``key`` or raise ValueError if it is unknown."""
    reg = _registry()
    if key not in reg:
        raise UnknownSourceError(
            f"Unknown source_db {key!r}. Available: {', '.join(sources())}."
        )
    return reg[key]
